contrib/unbound-cert-setup.sh

```diff
@@ -0,0 +1,141 @@
+        echo "${CTL_BASE}.key exists"
+else
+        echo "generating ${CTL_BASE}.key"
+        openssl ecparam -genkey -name secp384r1 > ${CTL_BASE}.key || error "could not gen ecdsa"
+fi
+
+# create self-signed cert CSR for server
+cat > ca_request.cfg <<EOCAConfig
+[req]
+prompt                 = no
+distinguished_name     = req_distinguished_name
+x509_extensions        = req_v3_extensions
+
+[req_distinguished_name]
+commonName             = unbound-ca
+
+[req_v3_extensions]
+subjectKeyIdentifier   = hash
+authorityKeyIdentifier = keyid:always,issuer:always
+basicConstraints       = CA:true
+EOCAConfig
+
+test -f ca_request.cfg || error "could not create ca_request.cfg"
+
+echo "create ${CA_BASE}.pem (self signed certificate)"
+openssl req -key "${CA_BASE}.key" -config ca_request.cfg  -new -x509 -days "${DAYS}" -out "${CA_BASE}.pem" || error "could not create ${CA_BASE}.pem"
+# create trusted usage pem
+openssl x509 -in "${CA_BASE}.pem" -addtrust serverAuth -out "${CA_BASE}_trust.pem"
+
+# create server cert CSR and sign it, piped
+cat > server_request.cfg <<EOServerConfig
+[req]
+prompt                 = no
+distinguished_name     = req_distinguished_name
+
+[req_distinguished_name]
+commonName             = unbound
+EOServerConfig
+
+cat > server_exts.cfg <<EOServerConfig
+subjectAltName         = @alt_names
+subjectKeyIdentifier   = hash
+authorityKeyIdentifier = keyid:always,issuer:always
+
+[alt_names]
+DNS.1                  = $(hostname)
+DNS.2                  = unbound
+DNS.3                  = localhost
+EOServerConfig
+
+test -f server_request.cfg || error "could not create server_request.cfg"
+
+echo "create ${SVR_BASE}.pem (signed server certificate)"
+openssl req -key "${SVR_BASE}.key" -config server_request.cfg -new | openssl x509 -req -days "${DAYS}" -CA "${CA_BASE}_trust.pem" -CAkey "${CA_BASE}.key" -CAcreateserial -"${HASH}" -extfile server_exts.cfg -out "${SVR_BASE}.pem"
+test -f ${SVR_BASE}.pem || error "could not create ${SVR_BASE}.pem"
+
+
+# create client cert CSR and sign it, piped
+cat > client_request.cfg <<EOCertConfig
+[req]
+prompt                 = no
+distinguished_name     = req_distinguished_name
+
+[req_distinguished_name]
+commonName             = unbound-control
+EOCertConfig
+
+test -f client_request.cfg || error "could not create client_request.cfg"
+
+echo "create ${CTL_BASE}.pem (signed client certificate)"
+openssl req -key "${CTL_BASE}.key" -config client_request.cfg -new | openssl x509 -req -days "${DAYS}" -CA "${CA_BASE}_trust.pem" -CAkey "${CA_BASE}.key" -CAcreateserial -"${HASH}" -out "${CTL_BASE}.pem"
+test -f "${CTL_BASE}.pem" || error "could not create ${CTL_BASE}.pem"
+
+# set desired permissions
+chmod 0640 "${CA_BASE}.key" "${SVR_BASE}.key" "${CTL_BASE}.key"
+chmod 0644 "${CA_BASE}.pem" "${SVR_BASE}.pem" "${CTL_BASE}.pem"
+
+# cleanup
+rm -f ca_request.cfg client_request.cfg server_request.cfg server_exts.cfg
+rm -f "${CA_BASE}_trust.pem" "${CA_BASE}_trust.srl"
```

Comment:
When is this .srl file created? I don't see it referenced anywhere else. And why remove the self-signed root?

Reply:
I think the `.srl` was a leftover from the original unbound-control certificate setup script. See [this line.])(https://github.com/NLnetLabs/unbound/blob/master/smallapp/unbound-control-setup.sh.in#L78)

As for this script, removing the `.srl` is fine because we set the following flag
> -CAcreateserial
> 
>     With this option and the -CA option the **CA serial number file** is created if it does not exist. A random number is generated, used for the certificate, and saved into the serial number file determined as described above.

Reply:
Aha, I can't get rid of `-CAcreateserial` because we need a unique serial for each certificate issued by the (self-signed) root certificate. Openssl uses the `.srl` file to keep track of the most recent serial number. The way SRE uses this script is that, upon building a new DNS server, this script is automatically run by config management and any extra setup files are thrown away. We're never going to issue new certificates after the initial batch on a DNS server because they're all ephemeral with a lifetime measured in hours, rather than days.